### PR TITLE
Add command to generate reference documentation

### LINKED ISSUE #5206: Generate documentation reference markdown by parsing source code
## Feature description

The reference documentation should be generated from source code, instead of manually maintained, to ensure it is always up-to-date.

This can be done using https://github.com/WordPress/phpdoc-parser, which produces JSON output. Turning the JSON output into templated markdown should then be relatively straight-forward.

This depends on https://github.com/ampproject/amp-wp/issues/5119 to make sure we clearly communicate what the scope of the public API of the plugin is.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

This PR adds the following command to generate reference documentation for the plugin:

```
Generates the reference documentation by parsing the source code.

Uses phpDocumentor and custom reflector from the WP DevHub parser plugin.

## OPTIONS

<source_folder>
: Path to the source folder that contains the source files to be parsed.

<destination_folder>
: Path to the destination folder where the output should be written to.

[--format=<format>]
: Output format to generate.
---
default: markdown
options:
  - json
  - markdown
---
```

It also adds the generated documentation itself, which can be accessed from the `/docs` folders and includes index files as entry points in the form of `README.md` files.

Fixes #5206 

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by schlessera** [comment]
I'll still have to battle PHPCS, it seems...

File: docs/src/Parser/FileReflector.php [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	/**
+	 * Check whether a given node is a filter.
+	 *
+	 * @param Node $node Node to check.
+	 *
+	 * @return bool Whether the node is a filter.
+	 */
+	protected function isFilter( Node $node ) {
+		// Ignore variable functions.
+		if ( 'Name' !== $node->name->getType() ) {
```
Comment (westonruter) [inline docs/src/Parser/FileReflector.php:228]: Should this not be `$node->getType()`?

image

File: docs/src/Parser/FunctionCallReflector.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+final class FunctionCallReflector extends BaseReflector {
+
+	/**
+	 * Returns the name for this Reflector instance.
+	 *
+	 * @return string
+	 */
+	public function getName() {
+		if ( isset( $this->node->namespacedName ) ) {
+			return '\\' . implode( '\\', $this->node->namespacedName->parts );
```
Comment (westonruter) [inline docs/src/Parser/FunctionCallReflector.php:34]: Is this expected?

image

File: docs/src/Parser/MethodCallReflector.php [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+		switch ( $class ) {
+			case '$this':
+			case 'self':
+				$namespace = (string) $this->called_in_class->getNamespace();
+				$namespace = ( 'global' !== $namespace ) ? $namespace . '\\' : '';
+				$class     = '\\' . $namespace . $this->called_in_class->getShortName();
+				break;
+			case 'parent':
+				$class = '\\' . $this->called_in_class->getNode()->extends->toString();
```
Comment (westonruter) [inline docs/src/Parser/MethodCallReflector.php:174]: image

**Review by westonruter — COMMENTED** [review]
Not sure these are issues but they were flagged in IntelliJ.

**Review by westonruter — APPROVED** [review]

**Pull request merged**